### PR TITLE
First step in the rewrite of Trajectory: introduce Forkable to manage forks and iterators.

### PR DESCRIPTION
**Pull request by pleroy**

### CODE REVIEW

File: physics/forkable.hpp [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+#include "geometry/named_quantities.hpp"
+
+namespace principia {
+
+using geometry::Instant;
+
+namespace physics {
+
+template<typename Tr4jectory>
+struct ForkableTraits;
```
Comment (eggrobin) [inline physics/forkable.hpp]: Comment.

File: physics/forkable.hpp [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+// This template represents a trajectory which is forkable and iterable.  It
+// uses CRTP to make sure that the return type in the concrete class doesn't
+// involve Forkable.
```
Comment (eggrobin) [inline physics/forkable.hpp]: Use words that mean things.

File: physics/forkable_body.hpp [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+#include "physics/forkable.hpp"
+
+namespace principia {
+namespace physics {
+
+template<typename Tr4jectory>
+not_null<Tr4jectory*>
+Forkable<Tr4jectory>::NewFork(Instant const & time) {
+  CHECK(timeline_find(time) != timeline_end() ||
+        (!is_root() && time == position_in_parent_children_->first))
```
Comment (eggrobin) [inline physics/forkable_body.hpp]: use ForkTime

File: physics/forkable.hpp [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+  // Null for a root.
+  Tr4jectory* parent_ = nullptr;
+
+  // TODO(phl): The following two iterators should be optional because we don't
+  // really have a good value for roots.
+
+  // This iterator is never at |end()|.
+  typename Children::const_iterator position_in_parent_children_;
+
+  // This iterator is never at |end()| if the parent's timeline is empty.
```
Comment (eggrobin) [inline physics/forkable.hpp]: wtf

File: physics/forkable_body.hpp [hunk 5]
```diff
@@ -0,0 +1,240 @@
+#include "forkable.hpp"
+#pragma once
```
Comment (eggrobin) [inline physics/forkable_body.hpp]: wtf

File: physics/forkable_body.hpp [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+namespace physics {
+
+template<typename Tr4jectory>
+not_null<Tr4jectory*>
+Forkable<Tr4jectory>::NewFork(Instant const & time) {
+  CHECK(timeline_find(time) != timeline_end() ||
+        (!is_root() && time == position_in_parent_children_->first))
+      << "NewFork at nonexistent time " << time;
+
+  // May be at |timeline_end()|.
```
Comment (eggrobin) [inline physics/forkable_body.hpp]: Document semantics

File: physics/forkable_body.hpp [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+  // Now set the members of the child object.
+  auto& child_forkable = child_it->second;
+  child_forkable.parent_ = that();
+  child_forkable.position_in_parent_children_ = child_it;
+  child_forkable.position_in_parent_timeline_ = timeline_it;
+
+  // Copy the tail of the trajectory in the child object.
+  if (timeline_it != timeline_end()) {
+    child_forkable.timeline_insert(++timeline_it, timeline_end());
+  }
```
Comment (eggrobin) [inline physics/forkable_body.hpp]: make an issue to remove this?

File: physics/forkable_body.hpp [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+
+template<typename Tr4jectory>
+bool Forkable<Tr4jectory>::is_root() const {
+  return parent_ == nullptr;
+}
+
+template<typename Tr4jectory>
+not_null<Tr4jectory const*>
+Forkable<Tr4jectory>::root() const {
+  Tr4jectory const* ancestor = that();
```
Comment (eggrobin) [inline physics/forkable_body.hpp]: not_null

File: physics/forkable_body.hpp [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+  if (is_root()) {
+    return nullptr;
+  } else {
+    return &position_in_parent_children_->first;
+  }
+}
+
+template<typename Tr4jectory>
+bool Forkable<Tr4jectory>::Iterator::operator==(
+    Iterator const & right) const {
```
Comment (eggrobin) [inline physics/forkable_body.hpp]: remove whitespace

File: physics/forkable_body.hpp [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+  bool const this_at_end = at_end();
+  bool const right_at_end = right.at_end();
+  if (this_at_end != right_at_end) {
+    return false;
+  } else if (this_at_end) {
+    return true;
```
Comment (eggrobin) [inline physics/forkable_body.hpp]: If they're from two different timelines, they differ (-- them).

File: physics/forkable_body.hpp [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+  CHECK(!at_end());
+
+  // Check if there is a next child in the ancestry.
+  auto ancestry_it = ancestry_.begin();
+  if (++ancestry_it != ancestry_.end()) {
+    // There is a next child.  See if we reached its fork time.
+    Instant const& current_time = ForkableTraits<Tr4jectory>::time(current_);
+    not_null<Tr4jectory const*> child = *ancestry_it;
+    Instant child_fork_time = child->position_in_parent_children_->first;
+    if (current_time == child_fork_time) {
```
Comment (eggrobin) [inline physics/forkable_body.hpp]: Maybe this test is wrong?

File: physics/forkable_body.hpp [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+  if (current_ == ancestor->timeline_begin()) {
+    CHECK_NOTNULL(ancestor->parent_);
+    // At the beginning of the first timeline.  Push the parent in front of the
+    // ancestry and set |current_| to the fork point.  If the timeline is empty,
+    // keep going until we find a non-empty one or the root.
+    do {
+      current_ = ancestor->position_in_parent_timeline_;
+      ancestor = ancestor->parent_;
+      ancestry_.push_front(ancestor);
+    } while (ancestor->timeline_empty() && ancestor->parent_ != nullptr);
```
Comment (eggrobin) [inline physics/forkable_body.hpp]: Apparently this test is wrong too?

File: physics/forkable_body.hpp [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+Forkable<Tr4jectory>::Wrap(
+    not_null<const Tr4jectory*> const ancestor,
+    TimelineConstIterator const position_in_ancestor_timeline) const {
+  Iterator iterator;
+
+  // Go up the ancestry chain until we find |ancestor| and set |current_| to
+  // |position_in_ancestor_timeline|.  The ancestry has |forkable|
+  // at the back, and the object containing |current_| at the front.  If
+  // |ancestor| is not found in the ancestry, stop at the root.
+  not_null<Tr4jectory const*> ancest0r = that();
```
Comment (eggrobin) [inline physics/forkable_body.hpp]: :sob:

File: physics/forkable_body.hpp [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+typename Forkable<Tr4jectory>::Iterator
+Forkable<Tr4jectory>::Wrap(
+    not_null<const Tr4jectory*> const ancestor,
+    TimelineConstIterator const position_in_ancestor_timeline) const {
+  Iterator iterator;
+
+  // Go up the ancestry chain until we find |ancestor| and set |current_| to
+  // |position_in_ancestor_timeline|.  The ancestry has |forkable|
+  // at the back, and the object containing |current_| at the front.  If
+  // |ancestor| is not found in the ancestry, stop at the root.
```
Comment (eggrobin) [inline physics/forkable_body.hpp]: check failure

File: physics/forkable.hpp (hunk 1 above)
Comment (pleroy) [inline physics/forkable.hpp]: Done.

File: physics/forkable.hpp (hunk 2 above)
Comment (pleroy) [inline physics/forkable.hpp]: Done.

File: physics/forkable_body.hpp (hunk 3 above)
Comment (pleroy) [inline physics/forkable_body.hpp]: Done.

File: physics/forkable.hpp (hunk 4 above)
Comment (pleroy) [inline physics/forkable.hpp]: Done.

File: physics/forkable_body.hpp (hunk 5 above)
Comment (pleroy) [inline physics/forkable_body.hpp]: Done.

File: physics/forkable_body.hpp (hunk 6 above)
Comment (pleroy) [inline physics/forkable_body.hpp]: Done.

File: physics/forkable_body.hpp (hunk 7 above)
Comment (pleroy) [inline physics/forkable_body.hpp]: Done.

File: physics/forkable_body.hpp (hunk 8 above)
Comment (pleroy) [inline physics/forkable_body.hpp]: Done.

File: physics/forkable_body.hpp (hunk 9 above)
Comment (pleroy) [inline physics/forkable_body.hpp]: Done.

File: physics/forkable_body.hpp (hunk 13 above)
Comment (pleroy) [inline physics/forkable_body.hpp]: Ack.

File: physics/forkable_body.hpp (hunk 14 above)
Comment (pleroy) [inline physics/forkable_body.hpp]: Done.

File: physics/forkable_body.hpp (hunk 11 above)
Comment (pleroy) [inline physics/forkable_body.hpp]: No, I think it works, added test IteratorIncrementMultipleForksSuccess to exercise the case that you were concerned about.

File: physics/forkable_body.hpp (hunk 12 above)
Comment (pleroy) [inline physics/forkable_body.hpp]: Yes, added test IteratorDecrementMultipleForksSuccess and fixed the bug.

File: physics/forkable_body.hpp (hunk 10 above)
Comment (pleroy) [inline physics/forkable_body.hpp]: Good point, added IteratorEndEquality and added code to normalize the ends.

**Pull request merged**